tile/codegen/cstr_reduction.cc

```diff
@@ -34,76 +39,169 @@ static void EvaluatePolynomial(const Polynomial<int64_t> orig_poly, const AliasM
+    for (auto kvp_it = constraint_map.begin(); kvp_it != constraint_map.end();) {
+      const auto& kvp = *kvp_it++;
       if (kvp.first == "") {
         continue;
       }
       Index* idx = block->idx_by_name(kvp.first);
-      if (idx->has_tags(skip_idxs)) {
+      if (idx->has_any_tags(skip_idxs)) {
+        IVLOG(4, "  Skipping index=" << *idx << " (skip_idxs=" << skip_idxs << ")");
         continue;
       }
-      int64_t coeff = -kvp.second;
-      if (idx->affine.getMap().empty() && coeff > 0) {
+      if (!idx->affine.getMap().empty()) {
+        IVLOG(4, "  Skipping non-constant index=" << *idx);
+      }
+      int64_t coeff = kvp.second;
+      IVLOG(4, "  Considering index=" << *idx << " with coeff=" << coeff);
+
+      // Check if we can reduce the index range by considering the constraint's lim value.
+      //
+      // Given an idx:
+      //     coeff*idx + (rest of the constraint) >= 0
+      // And assuming:
+      //     coeff < 0
+      // .. i.e. when the constraint is == lim_value-1, idx == 0
+      //    i.e. increasing idx only makes the overall constraint decrease.
+      // Then: coeff*idx >= -rest
+      //       -coeff*idx <= rest
+      //       idx <= rest / -coeff  (rounding division upwards -- N.B. -coeff is positive)
+      //       idx <= RoundUp(rest, -coeff)
+      // And since the maximum value (rest) might take on is lim_value-1,
+      //       idx <= RoundUp(lim_value-1, -coeff)
+      // So:
+      //       range'(idx) = min(range(idx), RoundUp(lim_value-1, -coeff) + 1)
+      if (coeff < 0) {
         uint64_t range = idx->range;
-        // Here max_value must be larger than 0, so we can cast it to unsigned
-        if (static_cast<uint64_t>(max_value) - 1 < (range - 1) * coeff) {
-          uint64_t new_range = static_cast<int>(Floor((max_value - 1) / coeff)) + 1;
+        // N.B. 0 < lim_value, or the block would've already been marked for removal.
+        uint64_t new_range = std::min(range, RoundUp(static_cast<uint64_t>(lim_value - 1), -coeff) + 1);
+        if (range != new_range) {
           IVLOG(4, "Reduce range of " << kvp.first << " from " << range << " to " << new_range);
           idx->range = new_range;
+          min_value += (range - new_range) * -coeff;
+          IVLOG(4, "  New min=" << min_value);
+        }
+      }
+
+      // Check if we can reduce the index range by considering the constraint's min value.
+      //
+      // Given an idx:
+      //     coeff*idx + (rest of the constraint) >= 0
+      // And assuming:
+      //     0 < coeff
+      // .. i.e. when the constraint is == min_value, idx == 0
+      //    i.e. increasing idx only makes the overall constraint increase.
+      // Then: coeff*idx >= -rest
+      //       idx >= -rest / coeff  (rounding division upwards -- N.B. coeff is positive)
+      //       idx >= RoundUp(-rest, coeff)
+      // And since the minimum value (rest) might take on is min_value,
+      //       idx >= RoundUp(-min_value, coeff)
```

Comment:
Something's wrong here but I don't know why. Will get back to you later.

Reply:
Fixed, I think.  It was actually a bit trickier to work through than what we were thinking, although the final equation is actually pretty clear (IMHO).